test/user_defined_output/log_response_headers_plugin_test.cc

```diff
@@ -0,0 +1,227 @@
+
+private:
+  std::vector<const Envoy::Http::HeaderEntry*> header_entries_;
+};
+
+/**
+ * Creates a LogResponseHeadersPlugin.
+ *
+ * @param config_textproto the textproto of the LogResponseHeadersConfig.
+ * @param fake_logger if not nullptr, injects this header logger to use in the created
+ * LogResponseHeadersPlugin.
+ * @return UserDefinedOutputPluginPtr
+ */
+UserDefinedOutputPluginPtr CreatePlugin(const std::string& config_textproto,
+                                        std::unique_ptr<HeaderLogger> header_logger) {
+  LogResponseHeadersConfig config;
+  TextFormat::ParseFromString(config_textproto, &config);
+
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  auto& factory = Envoy::Config::Utility::getAndCheckFactoryByName<UserDefinedOutputPluginFactory>(
+      "nighthawk.log_response_headers_plugin");
+  WorkerMetadata metadata;
+  metadata.worker_number = 1;
+
+  UserDefinedOutputPluginPtr plugin = factory.createUserDefinedOutputPlugin(config_any, metadata);
+  if (header_logger != nullptr) {
+    auto logging_plugin = dynamic_cast<LogResponseHeadersPlugin*>(plugin.get());
+    logging_plugin->injectHeaderLogger(std::move(header_logger));
+  }
+
+  return plugin;
+}
+
+/**
+ * Creates an empty LogResponseHeadersOutput.
+ */
+Envoy::ProtobufWkt::Any CreateOutput() {
+  LogResponseHeadersOutput output;
+
+  Envoy::ProtobufWkt::Any output_any;
+  output_any.PackFrom(output);
+
+  return output_any;
+}
+
+TEST(LogResponseHeadersPluginFactory, CreateEmptyConfigProtoCreatesCorrectType) {
+  auto& factory = Envoy::Config::Utility::getAndCheckFactoryByName<UserDefinedOutputPluginFactory>(
+      "nighthawk.log_response_headers_plugin");
+  Envoy::ProtobufTypes::MessagePtr empty_config = factory.createEmptyConfigProto();
+  LogResponseHeadersConfig expected_config;
+  EXPECT_THAT(*empty_config, EqualsProto(expected_config));
+}
+
+TEST(LogResponseHeadersPluginFactory, FactoryRegistersUnderCorrectName) {
+  LogResponseHeadersConfig config;
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  auto& factory = Envoy::Config::Utility::getAndCheckFactoryByName<UserDefinedOutputPluginFactory>(
+      "nighthawk.log_response_headers_plugin");
+  EXPECT_EQ(factory.name(), "nighthawk.log_response_headers_plugin");
+}
+
+TEST(LogResponseHeadersPluginFactory, CreateUserDefinedOutputPluginCreatesCorrectPluginType) {
+  LogResponseHeadersConfig config;
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  auto& factory = Envoy::Config::Utility::getAndCheckFactoryByName<UserDefinedOutputPluginFactory>(
+      "nighthawk.log_response_headers_plugin");
+  UserDefinedOutputPluginPtr plugin = factory.createUserDefinedOutputPlugin(config_any, {});
+  EXPECT_NE(dynamic_cast<LogResponseHeadersPlugin*>(plugin.get()), nullptr);
+}
+
+TEST(GetPerWorkerOutput, ReturnsProtoOfCorrectType) {
+  std::unique_ptr<HeaderLogger> logger = nullptr;
```

Comment:
Can you help me understand these two lines? Are we moving a nullptr into CreatePlugin? Is that relevant to this test case?

I would expect to satisfy the API with a valid logger instance when setting up, unless we are testing a case where nullptr is handled. But looking at the API, nullptr is not declared as a valid value.

If this is just a shortcut here, I would recommend to not take it so that future test readers aren't mislead.

Reply:
I had added a line to the private function CreatePlugin that for a nullptr it doesn't inject the test logger. It was intended as a way to test that the default Envoy logger work to some extent.

But I see how that adds confusion to readability on these tests, and so I've removed that and am now using the fake logger everywhere.